pkg/reconciler/managed/reconciler.go

```diff
@@ -675,6 +686,34 @@ func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error)
 			managed.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errReconcileCreate)))
 			return reconcile.Result{RequeueAfter: r.shortWait}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
 		}
+		log.Debug("Successfully requested creation of external resource", "requeue-after", time.Now().Add(r.shortWait))
+		record.Event(managed, event.Normal(reasonCreated, "Successfully requested creation of external resource"))
+
+		if creation.ResourceLateInitialized {
+			err := r.client.Update(ctx, managed)
+			if err != nil {
```

Comment:
I'd prefer to be able to avoid making superfluous calls to the external API if possible (i.e. create, delete, then create again). Do we have a good idea of why updates fail here? Is it due to write contention with the XR controller, i.e. having a stale resource version?

If we're seeing updates fail for some reason other than a stale resource version perhaps this is one niche case in which retrying the update inside the same reconcile loop would work; we could use exponential backoff with a small retry cap.

If we are seeing updates fail mostly due to stale resource versions things get trickier, but there are some options that don't require us to delete and recreate the resource in the provider's API, i.e.:

* Patch the resource after creation instead of updating it. This could be sufficient if we're confident that we'll only ever really be calling update to persist the resource's newly allocated external name annotation.
* The nuclear option; do something like the `APIUpdatingApplicator` does and retry the update with a manually overridden fresh resource version.
* Require that resources that generate their own IDs be somehow annotated by the Create call such that we could find them again on subsequent creates if they were orphaned? i.e. Tag VPCs with the `kube-system` namespace's UID and the name of the managed resource.
* Create some new type that acts as a temporary store of leaked external resources? Something like:

```yaml
apiVersion: crossplane.io/v1alpha1
kind: LeakedResource
metadata:
  name: example
managedResourceRef:
  apiVersion: ec2.crossplane.io/v1beta1
  kind: VPC
  name: example
externalName: vpc-a828k2dd9
```

Reply:
> If we are seeing updates fail mostly due to stale resource versions things get trickier.

I believe this is the case.

> The nuclear option; do something like the APIUpdatingApplicator does and retry the update with a manually overridden fresh resource version.

This sounds good to me. One risk is the case where we're actually stale, i.e. accidentally override some user input. We can use `APIUpdatingApplicator` to get the resource, change the annotation and let it update, i.e. do not touch any other thing meanwhile. Alternatively, we could do that via patching where the patch is manually curated in the code to have only the annotation so that we're sure that's the only thing we touch.

We can have an option in the applicator we use for exponential backoff so that we cover the case where api-server itself is down for a while (might be limited to 30 secs which is what we have in context for non-external-client operations).

> Require that resources that generate their own IDs be somehow annotated by the Create call such that we could find them again on subsequent creates if they were orphaned? i.e. Tag VPCs with the kube-system namespace's UID and the name of the managed resource.

I'm not sure if that's possible for all cases. A lot of resources have separate APIs for tagging, which could fail during creation of the resource and also brings in its own complications like what to do for cases where we want to import the resource.

> Create some new type that acts as a temporary store of leaked external resources?

I'm not sure. Nuke option seems simpler to manage to me.